_posts/2022-04-27-torch-outside-the-box/torch-outside-the-box.Rmd

```diff
@@ -0,0 +1,122 @@
+
+Even though scripting is not available from R (unless the scripted code is written in Python[^2]), we still benefit from its existence. When Python-side extension libraries use TorchScript (instead of normal C++ code), *we don't need to add bindings to the respective functions on the R (C++) side*. Instead, everything is taken care of by PyTorch.
+
+[^2]: Cf. the [TorchScript vignette](https://github.com/mlverse/torch/blob/main/vignettes/torchscript.Rmd) if you are interested in doing this.
+
+This -- although completely transparent to the user -- is what enables scenario one. In (Python) TorchVision, the pre-trained models provided will often make use of (model-dependent) special operators. Thanks to their having been scripted, we don't need to add a binding for each operator, let alone re-implement them on the R side.
+
+Having outlined some of the underlying functionality, we now present the scenarios themselves.
+
+## Scenario one: Load a TorchVision pre-trained model
+
+Perhaps you've already used one of the pre-trained models made available by TorchVision: A subset of these have been manually ported to `torchvision`, the R package. But there's more of them -- a *lot* more. Many use specialized operators, operators seldom needed outside of some algorithm's context. There would appear to be little use in creating R wrappers for those. And of course, continual appearance of new models would require continual porting efforts, on our side.
+
+Luckily, there is an elegant and effective solution. All the necessary infrastructure is set up by the lean, dedicated-purpose package [`torchvisionlib`](https://github.com/mlverse/torchvisionlib). (It can afford to be lean due to the Python side's liberal use of TorchScript, as explained in the previous section. But to the user -- whose perspective I'm taking in this scenario -- these details do not, of course, need to matter.)
+
+Once you've installed and loaded `torchvisionlib`, you have the choice among an impressive number of [image recognition-related models](https://pytorch.org/vision/stable/models.html). The process then is two-fold: You instantiate the model in Python, *script* it, and save it. Then, you load and use the model in R.
```

Comment:
I might actually list the two-fold process in an ordered list here for clarity (since the individual steps have sub-steps), e.g.

The process, then, is two-fold: 
1. You instantiate the model in Python, *script* it, and save it. 
2. You load and use the model in R.

Reply:
yes, that sounds good. Will do!